### ISSUE TITLE
LDtk Bug

### ISSUE THREAD

**Issue by Tsunami014** [deepnight/ldtk#970]
Steps to reproduce:

open LDtk
open [test.json](https://github.com/deepnight/ldtk/files/12861937/test.json)
get the bug

I don't know why. It is *probably* to do with the json file itself as I am using python to generate it based off a base file.

But it worked before (e.g. a couple of minutes ago) and now I have no clue what's happening. Maybe it's my json file, but then I thought it would catch an error to do with the file itself and present it differently... maybe you didn't catch this one

It did not update as I (was) offline

I'll attempt to debug the python and see what I can do to fix it.

Stack:
```
LDtk version: 1.4.1-64bits
Cannot read properties of undefined (reading 'pxHei')
TypeError
TypeError: Cannot read properties of undefined (reading 'pxHei')
    at data_inst_LayerInstance.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:42768:119)
    at data_Level.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:27991:10)
    at data_World.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:31814:6)
    at data_Project.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:29758:6)
    at page_Editor.selectProject (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206852:16)
    at new page_Editor (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206551:7)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3883:12
    at App.loadPage (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3867:25)
    at ui_ProjectLoader.onLoad (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3882:11)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:252163:11
```
Processes:
```
#0 App
 |--- #1 page.Home
 |--- #3 page.Editor
       |--- #4 misc.FileWatcher
       |--- #5 display.WorldRender
       |--- #6 display.LevelRender
       |--- #7 display.Camera
       |--- #8 display.Rulers[--]
       |--- #9 tool.SelectionTool[--]
       |--- #10 tool.lt.DoNothing[--]
       |--- #11 WorldTool
       |--- #12 tool.PanView[--]
#2 ui.modal.Progress
```

Log:
```


[BOOT]         App started
[BOOT]         Version: 1.4.1-64bits (build 471015)
[BOOT]         ExePath: C:/Users/Tsunami014/AppData/Local/Programs/ldtk
[BOOT]         Assets: C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets
[BOOT]         ExtraFiles: C:/Users/Tsunami014/AppData/Local/Programs/ldtk/extraFiles
[BOOT]         CWD: C:\Users\Tsunami014\AppData\Local\Programs\ldtk\
[BOOT]         Display: 1280x720
[BOOT]         Args: soloValues=[], args=[]
[FILE]         Loading settings from C:/Users/Tsunami014/AppData/Roaming/LDtk/settings...
[BOOT]         AppZoomFactor: 0.8
[BOOT]         Calling appReady...
[UPDATE]       Looking for update
[UPDATE]       ERROR: couldn't check for updates. Returned: net::ERR_INTERNET_DISCONNECTED
[UPDATE]       Fetching latest version from GitHub...
[ERROR]        Failed to fetch latest version from GitHub
[BOOT]         Start args: path=null levelIndex=null
[BOOT]         Loading Home...
[GENERAL]      Page started: page.Home()
[FILE]         Loading page template: home from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/home.html
[DEBUG]        samplesDir=C:/Users/Tsunami014/AppData/Local/Programs/ldtk/extraFiles/samples
[GENERAL]      Progress created.
[GENERAL]      Progress started: : "test.ldtk: Project...", 5 operation(s)
[LOAD]         Loading project test.ldtk...
[LOAD]           Project appBuildId=471015 appJsonVersion=1.4.1 jsonVersion=1.4.1
[LOAD]         Loading complete.
[GENERAL]      Page started: page.Editor()
[FILE]         Loading page template: editor from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/editor.html
[FILE]         Cleared all file watches
[GENERAL]      Page started: page.CrashReport()
[FILE]         Loading page template: crashReport from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/crashReport.html
[ERROR]        Cannot read properties of undefined (reading 'pxHei') (TypeError)
[ERROR]        TypeError: Cannot read properties of undefined (reading 'pxHei')
    at data_inst_LayerInstance.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:42768:119)
    at data_Level.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:27991:10)
    at data_World.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:31814:6)
    at data_Project.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:29758:6)
    at page_Editor.selectProject (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206852:16)
    at new page_Editor (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206551:7)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3883:12
    at App.loadPage (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3867:25)
    at ui_ProjectLoader.onLoad (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3882:11)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:252163:11

[GENERAL]      
#0 App
 |--- #1 page.Home
 |--- #3 page.Editor
       |--- #4 misc.FileWatcher
       |--- #5 display.WorldRender
       |--- #6 display.LevelRender
       |--- #7 display.Camera
       |--- #8 display.Rulers[--]
       |--- #9 tool.SelectionTool[--]
       |--- #10 tool.lt.DoNothing[--]
       |--- #11 WorldTool
       |--- #12 tool.PanView[--]
 |--- #13 page.CrashReport
#2 ui.modal.Progress

[FILE]         Cleared all file watches
```

**Comment by Tsunami014** [comment]
I just figured out something - this occurs when I do this:
```py
j = 0
for level in self.data['levels']:
    # other code
    for i in level['layerInstances']:
        i['levelId'] = j
        # other code
    j += 1
```
so without that code it creates the ldtk file normally---
but why I did that is because it duplicates 1 level (in the json) to become 2 and changes the intgrid of the second but in LDtk it still looks the same and I saw this and thought that's what I should change but apparently not.

But hang on now I look in the json schema and see `levelId: "Reference to the UID of the level containing this layer instance"`
But in the code I have already `level['uid'] = j` so why would `i['levelId'] = j` not work?

**Comment by Tsunami014** [comment]
wait I found out something else:
`i['layerDefUid'] = j` makes the whole thing blank even though the intgrids are stored in the levels AND also `level['uid'] = j`, which means that it's obviously attempting to render the intgrids on the wrong layer.
So that's my problem half solved. I just need to figure out what number to put the `i['layerDefUid']` for each layer.
But still some questions still remain, like what is a uid and why are some of these things I think should be working are not
I hope my randomising the iid to be random strings is not bad.

**Comment by Tsunami014** [comment]
I figured out a lot of things:
1. the setting the level uid to j was bad as that uid was already taken by other stuff, I now set it to nextUid and increase that by 1
2. layerDefUid *is* the uid of the layer
3. You set levelId to the uid of the level
I'm now doing `level['uid'] = nextUid`, `i['levelId'] = nextUid` and `nextUid += 1`, but it seems that whenever you change `levelId` It gives the error:

Stack:
```
LDtk version: 1.4.1-64bits
Cannot read properties of undefined (reading 'pxHei')
TypeError
TypeError: Cannot read properties of undefined (reading 'pxHei')
    at data_inst_LayerInstance.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:42768:119)
    at data_Level.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:27991:10)
    at data_World.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:31814:6)
    at data_Project.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:29758:6)
    at page_Editor.selectProject (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206852:16)
    at new page_Editor (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206551:7)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3883:12
    at App.loadPage (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3867:25)
    at ui_ProjectLoader.onLoad (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3882:11)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:252163:11
```
Processes:
```
#0 App
 |--- #20 page.Home
 |--- #23 page.Editor
       |--- #24 misc.FileWatcher
       |--- #25 display.WorldRender
       |--- #26 display.LevelRender
       |--- #27 display.Camera
       |--- #28 display.Rulers[--]
       |--- #29 tool.SelectionTool[--]
       |--- #30 tool.lt.DoNothing[--]
       |--- #31 WorldTool
       |--- #32 tool.PanView[--]
#22 ui.modal.Progress
```

Log:
```
[FILE]          -> Reload result: Ok
[FILE]          -> pixelData: Ok
[EVENT]        TilesetImageLoaded
[FILE]         Reloading tileset: Internal_Icons path=null
[FILE]         Loading atlas image: LdtkIcons...
[FILE]         Loading embed atlas LdtkIcons: C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/embedAtlas/finalbossblues-icons_full_16.png ...
[FILE]          -> Loaded 195310 bytes.
[FILE]          -> Decoded 512x1024 pixels.
[FILE]          -> Old size: 512x1024 -> 512x1024
[WARNING]      Tileset remapping (image size changed)...
[FILE]          -> Reload result: Ok
[FILE]          -> pixelData: Ok
[EVENT]        TilesetImageLoaded
[EVENT]        ProjectSelected
[RENDER]       Rendered level UI
[RENDER]       Rendering all world...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bounds...
[FILE]         Watching file: C:/Users/Tsunami014/Downloads/BlockCSS.png
[RENDER]       Rendering world level data.Level#8d9217c0-3b70-11ee-849e-1d317aac187d "Level_0"...
[RENDER]       Rendering world level fields data.Level#8d9217c0-3b70-11ee-849e-1d317aac187d "Level_0"...
[RENDER]       Rendering world bg...
[GENERAL]      Progress completed: "blank.ldtk: Project..." (1.2s)
[RENDER]       Rendering world level data.Level#2ed1b300-6280-11ee-adba-f7f488fbd48d "Level_1"...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bg...
[RENDER]       Rendering world bg...
[GENERAL]      Page started: page.Home()
[FILE]         Loading page template: home from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/home.html
[DEBUG]        samplesDir=C:/Users/Tsunami014/AppData/Local/Programs/ldtk/extraFiles/samples
[FILE]         Cleared all file watches
[GENERAL]      Progress created.
[GENERAL]      Progress started: : "test.ldtk: Project...", 5 operation(s)
[LOAD]         Loading project test.ldtk...
[LOAD]           Project appBuildId=471015 appJsonVersion=1.4.1 jsonVersion=1.4.1
[LOAD]         Loading complete.
[GENERAL]      Page started: page.Editor()
[FILE]         Loading page template: editor from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/editor.html
[FILE]         Cleared all file watches
[GENERAL]      Page started: page.CrashReport()
[FILE]         Loading page template: crashReport from C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/tpl/pages/crashReport.html
[ERROR]        Cannot read properties of undefined (reading 'pxHei') (TypeError)
[ERROR]        TypeError: Cannot read properties of undefined (reading 'pxHei')
    at data_inst_LayerInstance.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:42768:119)
    at data_Level.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:27991:10)
    at data_World.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:31814:6)
    at data_Project.tidy (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:29758:6)
    at page_Editor.selectProject (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206852:16)
    at new page_Editor (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:206551:7)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3883:12
    at App.loadPage (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3867:25)
    at ui_ProjectLoader.onLoad (file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:3882:11)
    at file:///C:/Users/Tsunami014/AppData/Local/Programs/ldtk/resources/app.asar/assets/js/renderer.js:252163:11

[GENERAL]      
#0 App
 |--- #20 page.Home
 |--- #23 page.Editor
       |--- #24 misc.FileWatcher
       |--- #25 display.WorldRender
       |--- #26 display.LevelRender
       |--- #27 display.Camera
       |--- #28 display.Rulers[--]
       |--- #29 tool.SelectionTool[--]
       |--- #30 tool.lt.DoNothing[--]
       |--- #31 WorldTool
       |--- #32 tool.PanView[--]
 |--- #33 page.CrashReport
#22 ui.modal.Progress

[FILE]         Cleared all file watches
```

**Comment by deepnight** [comment]
Hi!
A more reliable way to access a level is to use the `iid` (String) field which is guaranteed to be globlally unique, while `uid` is unique in the context only (eg. unique among Levels, but my be equal to an Entity uid).

Anyway, the current file is invalid because Level_0 & Level_1 share the same uid, which is 0. Just make sure any uid is context-unique and it should be fine :)

**Comment by deepnight** [comment]
I tested it, changed Level_1 uid to something unique (eg. 1), and the file loaded just right 👍

**Comment by Tsunami014** [comment]
Now I think it is my code's fault and it's very confusing rn and thanks for officially being the fastest non-discord reply I've ever gotten I'll do some work and see what I can do

**Comment by Tsunami014** [comment]
It WAS my code that was wrong.
1. I commented out `i['levelId'] = nextUid`. It still works. I don't care very much about debugging working things
2. previously I was using `{}.copy()` to copy a dict in python and it looked the same which sent me on this whole endeavour of attempting to change the levelId thing. I made use of `copy.deepcopy` and now it works :) :)
3. You were right, using nextUid or not made no difference.
4. It is still very confusing why it doesn't need the levelId to work and if you do add it it fails. But who knows...

So everything is cool now 😃
